### PR TITLE
bpo-40854: Allow overriding sys.platlibdir via PYTHONPLATLIBDIR env-var

### PR DESCRIPTION
**Pull request by manisandro**
You can currently point the python interpreter to a different install say via

    export PYTHONHOME=<Prefix>
    export PYTHONPATH=<Prefix>/lib/python3.9
    python3

With the newly added platlibdir [1], if python was configured with platlibdir=lib64, this will break because i.e. the site-packages dir as returned by `sysconfig.get_paths()` will use lib64 and not lib as the other install may be using.

This PR adds the possibility to override the platlibdir via environment variable.

Full rationale: [2].

[1] https://github.com/python/cpython/pull/8068
[2] https://src.fedoraproject.org/rpms/python3.9/pull-request/10

https://bugs.python.org/issue40854

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@manisandro

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Python/sysmodule.c [hunk 1]
```diff
@@ -2923,7 +2923,11 @@ _PySys_InitMain(PyThreadState *tstate)
     SET_SYS_FROM_WSTR("exec_prefix", config->exec_prefix);
     SET_SYS_FROM_WSTR("base_exec_prefix", config->base_exec_prefix);
     {
-        PyObject *str = PyUnicode_FromString(PLATLIBDIR);
+        const char* platlibdir = getenv("_PYTHON_PLATLIBDIR");
```
Comment (vstinner) [inline Python/sysmodule.c]: Why "_PYTHON_PLATLIBDIR" name, rather than PYTHONPLATLIBDIR?

You should use _Py_GetEnv(config->use_environment, "PYTHONPLATLIBDIR") if your environment variable name  starts with "PYTHON".

File: Python/sysmodule.c [hunk 2]
```diff
@@ -2923,7 +2923,11 @@ _PySys_InitMain(PyThreadState *tstate)
...
+        if(!platlibdir) {
+            platlibdir = PLATLIBDIR;
+        }
+        PyObject *str = PyUnicode_FromString(platlibdir);
```
Comment (vstinner) [inline Python/sysmodule.c]: It's wrong to decode an environment variable from UTF-8. You must use PyUnicode_DecodeFSDefault() for the env var, but keep PyUnicode_FromString for PLATLIBDIR.

**Review by vstinner — CHANGES_REQUESTED** [review]
See https://pythondev.readthedocs.io/pyconfig.html#add-an-environment-variable for the different steps to add a new environment variables:

* Update usage_xxx to document the new env var
* Doc/using/cmdline.rst: Document the new env var
* Misc/python.man: Document the new env var
* You may also mention it at https://docs.python.org/dev/library/sys.html#sys.platlibdir

**Comment by manisandro** [comment]
@vstinner Thanks for your prompt feedback, I've reworked the PR

**Comment by hroncok** [comment]
@manisandro Could you please copy paste and use the phrase as suggested by the bot? It will label this PR accordingly.

**Comment by manisandro** [comment]
I have made the requested changes; please review again

File: Doc/using/cmdline.rst [hunk 3]
```diff
@@ -538,6 +538,12 @@ conflict.
    within a Python program as the variable :data:`sys.path`.
 
 
+.. envvar:: PYTHONPLATLIBDIR
+
+   Override the python platform libdir name from the value set by
+   ``--with-platlibdir`` when the Python build was configured.
+
```
Comment (vstinner) [inline Doc/using/cmdline.rst:547]: Please add ".. versionadded:: 3.10".

File: Doc/using/cmdline.rst [hunk 4]
```diff
@@ -538,6 +538,12 @@ conflict.
    within a Python program as the variable :data:`sys.path`.
 
 
+.. envvar:: PYTHONPLATLIBDIR
+
+   Override the python platform libdir name from the value set by
+   ``--with-platlibdir`` when the Python build was configured.
```
Comment (vstinner) [inline Doc/using/cmdline.rst]:
```suggestion
   If this is set to a non-empty string, it overrides :data:`sys.platlibdir` value.
```

File: Programs/_testembed.c [hunk 5]
```diff
@@ -548,6 +548,11 @@ static int test_init_from_config(void)
     /* FIXME: test home */
     /* FIXME: test path config: module_search_path .. dll_path */
 
+    putenv("PYTHONPLATLIBDIR="PLATLIBDIR);
+    size_t len;
+    config_set_string(&config, &config.platlibdir, Py_DecodeLocale(PLATLIBDIR, &len));
```
Comment (vstinner) [inline Programs/_testembed.c]:
```
    putenv("PYTHONPLATLIBDIR=env_platlibdir"); 
    PyStatus status = PyConfig_SetBytesString(&config, &config.platlibdir, PLATLIBDIR);
    if (PyStatus_Exception(status)) {
        PyConfig_Clear(config);
        Py_ExitStatusException(status);
    }
```

File: Programs/_testembed.c [hunk 6]
```diff
@@ -668,6 +673,7 @@ static void set_most_env_vars(void)
     putenv("PYTHONFAULTHANDLER=1");
     putenv("PYTHONIOENCODING=iso8859-1:replace");
     putenv("PYTHONOLDPARSER=1");
+    putenv("PYTHONPLATLIBDIR="PLATLIBDIR);
```
Comment (vstinner) [inline Programs/_testembed.c]:
```suggestion
    putenv("PYTHONPLATLIBDIR=env_platlibdir"); 
```

File: Python/initconfig.c [hunk 7]
```diff
@@ -110,6 +110,7 @@ PYTHONPATH   : '%lc'-separated list of directories prefixed to the\n\
 static const char usage_5[] =
 "PYTHONHOME   : alternate <prefix> directory (or <prefix>%lc<exec_prefix>).\n"
 "               The default module search path uses %s.\n"
+"PYTHONPLATLIBDIR : alternate libdir name for platform libraries.\n"
```
Comment (vstinner) [inline Python/initconfig.c]:
```suggestion
"PYTHONPLATLIBDIR : override sys.platlibdir.\n"
```

File: Python/initconfig.c [hunk 8]
```diff
@@ -151,7 +152,7 @@ int Py_OptimizeFlag = 0; /* Needed by compile.c */
 int Py_NoSiteFlag = 0; /* Suppress 'import site' */
 int Py_BytesWarningFlag = 0; /* Warn on str(bytes) and str(buffer) */
 int Py_FrozenFlag = 0; /* Needed by getpath.c */
-int Py_IgnoreEnvironmentFlag = 0; /* e.g. PYTHONPATH, PYTHONHOME */
+int Py_IgnoreEnvironmentFlag = 0; /* e.g. PYTHONPATH, PYTHONHOME, PYTHONPLATLIBDIR */
```
Comment (vstinner) [inline Python/initconfig.c]: I don't think that it's useful to list all env vars. I suggest to revert this change.

File: Python/initconfig.c [hunk 9]
```diff
@@ -1235,7 +1239,6 @@ config_init_home(PyConfig *config)
                               L"PYTHONHOME", "PYTHONHOME");
 }
 
-
```
Comment (vstinner) [inline Python/initconfig.c]: Please revert this unrelated change. Also, I like two empty newlines.

File: Python/initconfig.c [hunk 10]
```diff
@@ -1731,6 +1742,11 @@ config_read(PyConfig *config)
         }
     }
 
+    if(config->platlibdir == NULL) {
+        size_t len;
+        config->platlibdir = Py_DecodeLocale(PLATLIBDIR, &len);
```
Comment (vstinner) [inline Python/initconfig.c]: You must handle errors. Example:

```
        status = CONFIG_SET_BYTES_STR(config, &config->program_name, p,
                                      "PYTHONEXECUTABLE environment variable");
        if (_PyStatus_EXCEPTION(status)) {
            return status;
        }
```

File: Misc/NEWS.d/next/Core and Builtins/2020-06-03-13-53-24.bpo-40854.O6vfQU.rst [hunk 11]
```diff
@@ -0,0 +1 @@
+Allow overriding sys.platlibdir via _PYTHON_PLATLIBDIR env-var
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2020-06-03-13-53-24.bpo-40854.O6vfQU.rst]:
```suggestion
Allow overriding :data:`sys.platlibdir` via a new :envvar:`PYTHONPLATLIBDIR` environment variable.
```

**Review by vstinner — CHANGES_REQUESTED** [review]
Thanks, the update PR is much better! New review.

Please document also the new PyConfig field in https://docs.python.org/dev/c-api/init_config.html#c.PyConfig documentation with a ".. versionadded:: 3.10" markup.

**Comment by manisandro** [comment]
I have made the requested changes; please review again

**Comment by vstinner** [comment]
It seems possible to test a custom PYTHONPATHLIBDIR in test_embed. At least, the following patch worked for me:

```
diff --git a/Lib/test/test_embed.py b/Lib/test/test_embed.py
index 3b551e9015..f1371db866 100644
--- a/Lib/test/test_embed.py
+++ b/Lib/test/test_embed.py
@@ -586,13 +586,14 @@ class InitConfigTests(EmbeddingTestsMixin, unittest.TestCase):
             if value is self.GET_DEFAULT_CONFIG:
                 expected[key] = config[key]
 
-        pythonpath_env = expected['pythonpath_env']
-        if pythonpath_env is not None:
-            paths = pythonpath_env.split(os.path.pathsep)
-            expected['module_search_paths'] = [*paths, *expected['module_search_paths']]
-        if modify_path_cb is not None:
-            expected['module_search_paths'] = expected['module_search_paths'].copy()
-            modify_path_cb(expected['module_search_paths'])
+        if expected['module_search_paths'] is not self.IGNORE_CONFIG:
+            pythonpath_env = expected['pythonpath_env']
+            if pythonpath_env is not None:
+                paths = pythonpath_env.split(os.path.pathsep)
+                expected['module_search_paths'] = [*paths, *expected['module_search_paths']]
+            if modify_path_cb is not None:
+                expected['module_search_paths'] = expected['module_search_paths'].copy()
+                modify_path_cb(expected['module_search_paths'])
 
         for key in self.COPY_PRE_CONFIG:
             if key not in expected_preconfig:
@@ -765,6 +766,8 @@ class InitConfigTests(EmbeddingTestsMixin, unittest.TestCase):
             'buffered_stdio': 0,
             'user_site_directory': 0,
             'faulthandler': 1,
+            'platlibdir': 'my_platlibdir',
+            'module_search_paths': self.IGNORE_CONFIG,
 
             'check_hash_pycs_mode': 'always',
             'pathconfig_warnings': 0,
@@ -796,6 +799,8 @@ class InitConfigTests(EmbeddingTestsMixin, unittest.TestCase):
             'user_site_directory': 0,
             'faulthandler': 1,
             'warnoptions': ['EnvVar'],
+            'platlibdir': 'env_platlibdir',
+            'module_search_paths': self.IGNORE_CONFIG,
             '_use_peg_parser': 0,
         }
         self.check_all_configs("test_init_compat_env", config, preconfig,
@@ -824,6 +829,8 @@ class InitConfigTests(EmbeddingTestsMixin, unittest.TestCase):
             'user_site_directory': 0,
             'faulthandler': 1,
             'warnoptions': ['EnvVar'],
+            'platlibdir': 'env_platlibdir',
+            'module_search_paths': self.IGNORE_CONFIG,
             '_use_peg_parser': 0,
         }
         self.check_all_configs("test_init_python_env", config, preconfig,
diff --git a/Makefile.pre.in b/Makefile.pre.in
index 1d0bd32ffb..a7d5dd2945 100644
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -782,9 +782,7 @@ Programs/python.o: $(srcdir)/Programs/python.c
 	$(MAINCC) -c $(PY_CORE_CFLAGS) -o $@ $(srcdir)/Programs/python.c
 
 Programs/_testembed.o: $(srcdir)/Programs/_testembed.c
-	$(MAINCC) -c $(PY_CORE_CFLAGS) \
-		-DPLATLIBDIR='"$(PLATLIBDIR)"' \
-		-o $@ $(srcdir)/Programs/_testembed.c
+	$(MAINCC) -c $(PY_CORE_CFLAGS) -o $@ $(srcdir)/Programs/_testembed.c
 
 Modules/_sre.o: $(srcdir)/Modules/_sre.c $(srcdir)/Modules/sre.h $(srcdir)/Modules/sre_constants.h $(srcdir)/Modules/sre_lib.h
 
diff --git a/Programs/_testembed.c b/Programs/_testembed.c
index c6a0d6c9e0..7621c46260 100644
--- a/Programs/_testembed.c
+++ b/Programs/_testembed.c
@@ -549,7 +549,7 @@ static int test_init_from_config(void)
     /* FIXME: test path config: module_search_path .. dll_path */
 
     putenv("PYTHONPLATLIBDIR=env_platlibdir");
-    status = PyConfig_SetBytesString(&config, &config.platlibdir, PLATLIBDIR);
+    status = PyConfig_SetBytesString(&config, &config.platlibdir, "my_platlibdir");
     if (PyStatus_Exception(status)) {
         PyConfig_Clear(&config);
         Py_ExitStatusException(status);
@@ -676,7 +676,7 @@ static void set_most_env_vars(void)
     putenv("PYTHONFAULTHANDLER=1");
     putenv("PYTHONIOENCODING=iso8859-1:replace");
     putenv("PYTHONOLDPARSER=1");
-    putenv("PYTHONPLATLIBDIR="PLATLIBDIR);
+    putenv("PYTHONPLATLIBDIR=env_platlibdir");
 }
 
 
```

**Comment by manisandro** [comment]
Thanks, added to PR.

File: Doc/c-api/init_config.rst [hunk 12]
```diff
@@ -436,6 +436,14 @@ PyConfig
 
       :data:`sys.base_prefix`.
 
+   .. c:member:: wchar_t* platlibdir
+
+      :data:`sys.platlibdir`: platform library dir name, set at configure time
```
Comment (vstinner) [inline Doc/c-api/init_config.rst]:
```suggestion
      :data:`sys.platlibdir`: platform library directory name, set at configure time
```

File: Misc/python.man [hunk 13]
```diff
@@ -413,6 +413,9 @@ inserted in the path in front of $PYTHONPATH.
 The search path can be manipulated from within a Python program as the
 variable
 .IR sys.path .
+.IP PYTHONPLATLIBDIR
+Override the python platform libdir name from the value set by
+--with-platlibdir when the Python build was configured.
```
Comment (vstinner) [inline Misc/python.man]: Override sys.platlibdir.

File: Programs/_testembed.c [hunk 14]
```diff
@@ -548,6 +548,14 @@ static int test_init_from_config(void)
     /* FIXME: test home */
     /* FIXME: test path config: module_search_path .. dll_path */
 
+    putenv("PYTHONPLATLIBDIR=env_platlibdir");
+    status = PyConfig_SetBytesString(&config, &config.platlibdir, "my_platlibdir");
+    if (PyStatus_Exception(status)) {
+        PyConfig_Clear(&config);
+        Py_ExitStatusException(status);
+    }
+
```
Comment (vstinner) [inline Programs/_testembed.c:557]: nitpick: one empty line is enough :-)

**Review by vstinner — COMMENTED** [review]
Ok, the PR now looks complete: all documentations are updated, thanks!

I didn't expect the new sys.platlibdir to land into PyConfig, but I think that I understood the rationale in https://bugs.python.org/issue40854

Overriding PYTHONHOME is uncommon. If you do that, I'm not surprised that you fall into corner cases. So having the ability to override platlibdir sounds reasonable.

Hum. It seems like PyConfig.platlibdir is a parameter which affects the "Path Configuration". So please add it to "Path configuration inputs" at:
https://docs.python.org/dev/c-api/init_config.html#path-configuration

**Comment by manisandro** [comment]
Thanks, I've updated the PR accordingly.

File: Python/initconfig.c [hunk 15]
```diff
@@ -1731,6 +1743,14 @@ config_read(PyConfig *config)
         }
     }
 
+    if(config->platlibdir == NULL) {
+        status = CONFIG_SET_BYTES_STR(config, &config->platlibdir, PLATLIBDIR,
+                                      "PLATLIBDIR macro");
```
Comment (vstinner) [inline Python/initconfig.c:1752]: Oh, I didn't notice the Makefile change to define PLATLIBDIR macro. Please add an explicit check to ensure that it's defined. See Modules/getpath.c. I sugges to add at the top of this file, after includes (on Windows, it's defined by PC/pyconfig.h which is included by Python.h):

```
#ifndef PLATLIBDIR
#  error "PLATLIBDIR macro must be defined"
#endif
```

**Review by vstinner — COMMENTED** [review]
Oh, one last request, sorry!

When you update a PR, please add a new commit rather than always rebase/squash your commits. New commits help to see what has been modified since my latest review.

**Comment by manisandro** [comment]
No problem, pushed as new commit.

**Comment by vstinner** [comment]
Thanks, I merged your PR.

**Comment by manisandro** [comment]
Thank you for merging and for the review!
